File: Python/ByTier/C/2110_C_Racing.py
# Problem: CF 2110 C - Racing
# https://codeforces.com/contest/2110/problem/C

"""
C. Racing

Purpose:
This code solves a problem where a drone must navigate through a sequence of obstacles,
each with a height constraint [l_i, r_i]. The drone's movement between obstacles is defined 
by a difference array `d`, where each element can be 0 or 1 (or -1 for unknown). 
The task is to assign values to the -1 positions such that the cumulative height stays 
within all obstacle constraints.

Algorithms/Techniques:
- Two-pointer sliding window technique with stack-based tracking of -1 elements
- Greedy assignment of 0s and 1s while maintaining bounds

Time Complexity: O(n) per test case, where n is the number of obstacles.
Space Complexity: O(n) for storing the difference array and auxiliary stack.

Input:
Each test case consists of:
    - n (number of obstacles)
    - Array d of size n with values in {-1, 0, 1}
    - For each obstacle i, two integers l_i and r_i defining valid height range

Output:
For each case, output the completed array d or -1 if impossible.
"""

import sys

input = sys.stdin.readline
for i in range(int(input())):
    n = int(input())
    a = list(map(int, input().split()))
    s = []  # Stack to store indices of -1 elements
    u, v, f = 0, 0, 1  # u: count of 1s, v: current height, f: flag for feasibility
    
    for j in range(n):
        l, r = map(int, input().split())
        if a[j] == -1:
            s.append(j)  # Track index of unknown value
            v += 1       # Assume we climb by 1 to reach max height at that point
        elif a[j] == 1:
            u += 1       # Count of 1s increases 
            v += 1       # Height also climbs by 1
            
        # If current height exceeds upper bound, we must reduce it by changing some 1s to 0s
        while v > r and s:
            a[s.pop()] = 0   # Change 1s in stack to 0 (climb down)
            v -= 1           # Reduce height accordingly
            
        # If count of 1s is below lower bound, we must change 0s to 1s
        while u < l and s:
            a[s.pop()] = 1   # Change 0 to 1 to increase count of 1s
            u += 1           # Increase number of climbs
            
        # Check if current height or number of climbs violates constraints
        if v < u or v > r or u < l:
            f = 0    # Set flag to indicate failure
            
    # After processing, reset remaining -1s to 0 to ensure final heights are valid
    for j in range(n):
        if a[j] == -1:
            a[j] = 0
            
    # Print result or -1 if impossible
    if f:
        print(*a)
    else:
        print(-1)


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/C/2110_C_Racing.cpp
// Problem: CF 2110 C - Racing
// https://codeforces.com/contest/2110/problem/C

/*
C. Racing
time limit per test2 seconds
memory limit per test256 megabytes

Algorithm: Greedy with Stack
Approach:
- Process obstacles sequentially maintaining current height `h` and a stack of indices where `d[i] = -1`
- When encountering a `-1`, push index to stack (unknown step)
- When encountering a positive `d[i]`, increase height by that amount
- When insufficient height, greedily assign 1s to previous unknowns in the stack to make up the difference
- Ensure height stays within bounds [bl[p], bh[p]] at all times
- After processing, pop remaining stack elements and assign them 0
Time Complexity: O(n) - each element is pushed/poppped from stack at most once
Space Complexity: O(n) - for storing input data, stack

*/

#include <cstdio>
#include <stack>
#include <vector>
int main() {
  long t;
  scanf("%ld", &t);
  while (t--) {
    long n;
    scanf("%ld", &n);
    std::vector<long> dv(n), bl(n), bh(n);
    for (long p = 0; p < n; p++) {
      scanf("%ld", &dv[p]);
    }
    for (long p = 0; p < n; p++) {
      scanf("%ld %ld", &bl[p], &bh[p]);
    }
    bool possible(true);
    std::stack<long> s;
    long h(0);
    for (long p = 0; possible && p < n; p++) {
      if (dv[p] < 0) { // If unknown step, push index to stack
        s.push(p);
      } else {
        h += dv[p]; // Add known height change to current height
      }
      while (!s.empty() && h < bl[p]) { // Not enough height to pass obstacle
        dv[s.top()] = 1; // Assign upward step to make up for deficit
        ++h;
        s.pop();
      }
      if (h < bl[p] || h > bh[p]) { // Height out of valid range
        possible = false;
        break;
      }
      while (h + s.size() > bh[p]) { // Too much potential height increase from stack
        dv[s.top()] = 0; // Assign no step to prevent overflow
        s.pop();
      }
    }
    if (!possible) {
      puts("-1");
      continue;
    }
    while (!s.empty()) { // Assign remaining unknowns as 0 (no steps)
      dv[s.top()] = 0;
      s.pop();
    }
    for (long p = 0; p < n; p++) {
      printf("%ld ", dv[p]);
    }
    puts("");
  }
}


// https://github.com/VaHiX/codeForces/